### PR TITLE
Register a new last_id once per slot

### PR DESCRIPTION
**Pull request by mvines**
This PR moves last_id generation from once per tick to once per slot (the last tick in a slot is registered as a last_id).

TODO:
* [x] Clean up, fix up some tests

### CODE REVIEW

File: runtime/src/bank.rs [hunk 1]
```diff
@@ -97,6 +97,9 @@ pub struct Bank {
     /// Hash of this Bank's parent's state
     parent_hash: Hash,
 
+    /// Bank tick height
+    tick_height: AtomicUsize, // TODO: Use AtomicU64 if/when available
```
Comment (mvines) [inline runtime/src/bank.rs]: 🤢

File: runtime/src/bank.rs [hunk 2]
```diff
@@ -354,15 +359,23 @@ impl Bank {
...
         // assert!(!self.is_frozen());
+
         let current_tick_height = {
-            // Atomic register and read the tick
+            self.tick_height.fetch_add(1, Ordering::SeqCst);
+            self.tick_height.load(Ordering::SeqCst) as u64
+        };
+
+        // Register a new block hash if this is the last tick in the slot
+        if self.tick_index() == self.ticks_per_slot - 1 {
```
Comment (mvines) [inline runtime/src/bank.rs]: Here's where we change to 1 last_id per slot

File: runtime/src/bank.rs [hunk 3]
```diff
@@ -743,7 +756,11 @@ impl Bank {
 
     /// Return the number of ticks since genesis.
     pub fn tick_height(&self) -> u64 {
-        self.tick_hash_queue.read().unwrap().hash_height()
+        // tick_height is using an AtomicUSize because AtomicU64 is not yet a stable API.
+        // Until we can switch to AtomicU64, fail if usize is not the same as u64
+        assert_eq!(std::usize::MAX, 0xFFFFFFFFFFFFFFFF);
```
Comment (mvines) [inline runtime/src/bank.rs]: Any better ideas?

**Comment by garious** [comment]
One complication, I think this needs to be the last tick hash of each slot, not the block hash. What we're doing with this queue is attempting to guarantee that a transaction will be processed if received within a certain amount of *time*. Once that hash is purged, the leader should reject it.  Using only block hash makes that queue a little bit longer (if slots are skipped), and I'm not sure that's okay.

**Comment by mvines** [comment]
I think it already is the last tick in the slot though.

Right now hash_queue attempts to keep 2 minutes worth of tick hashes.  While not in this PR yet, I was planning to adjust the constants so that we end up with 2 minutes worth of block hashes instead.  So same time window from the client’s pov

**Comment by mvines** [comment]
@garious 🛎 

I think all your previous concerns are addressed:
1. The last tick hash of a slot is registered as a `last_id` (which I'll globally rename to `last_block_hash` after this PR lands)
2. The time window for valid `last_id`s remains the same (based on the the value of `solana_sdk::timing::MAX_RECENT_BLOCK_HASHES`)

File: src/blocktree_processor.rs [hunk 4]
```diff
@@ -439,19 +447,20 @@ mod tests {
...
+        solana_logger::setup();
         let leader_pubkey = Keypair::new().pubkey();
         let (genesis_block, mint_keypair) = GenesisBlock::new_with_leader(100, leader_pubkey, 50);
-        let (ledger_path, last_id) = create_new_tmp_ledger!(&genesis_block);
+        let (ledger_path, mut last_entry_hash) = create_new_tmp_ledger!(&genesis_block);
         debug!("ledger_path: {:?}", ledger_path);
 
         let mut entries = vec![];
-        let mut last_entry_hash = last_id;
+        let last_id = genesis_block.hash();
```
Comment (garious) [inline src/blocktree_processor.rs]: I'm a bit confused on the few places in this PR that still have `last_id`. Rebase error?

Comment (mvines) [reply src/blocktree_processor.rs]: Oh.  I'll do the full `last_id` renaming in a separate PR, that'll be a mechanical change that nobody wants to review.   This PR is the 🍖 that actually needs review, so tried to keep it as small and tight as possible

**Comment by garious** [comment]
LGTM. It's pretty interesting that in blocktree there's zero or one block per slot, but when we process it, there's exactly one block per slot. That is, when a block covers two slots, the HashQueue treats it as two blocks.  This seems exactly right to me, but it's surprising. I wonder how long before someone trips up on it.

**Pull request merged**